android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/INIParser.java

```diff
@@ -0,0 +1,123 @@
+		for(int i=0; i < linesArray.length; i++) {
+			String line = linesArray[i].trim();
+			if(!(line.equals("") || line.charAt(0) == ';')) {
+				lines.add(line);
+			}
+		}
+		
+		// Allow a plain default section - if it goes unutilized, we can remove it later.
+		String currentSectionName = "";
+		Map<String, List<String>> currentSection = new HashMap<String, List<String>>();
+		sections.put(currentSectionName, currentSection);
+		
+		// Main parsing loop.
+		for(String line:lines) {
+			if(line.charAt(0) == '[') {
+				currentSectionName = unquote_helper(line, '[', ']', 
+						"A section definition is missing its closing bracket",
+						"Improper section header definition - extra characters detected");
+						
+				
+				if(sections.containsKey(currentSectionName)) {
+					throw new INIParseException("Section \"" + currentSectionName + "\" has a duplicate entry!");
+				}
+				
+				currentSection = new HashMap<String, List<String>>();
+				sections.put(currentSectionName, currentSection);
+			} else if(line.indexOf('=') != -1) {
+				int eq = line.indexOf('=');
+				
+				String key = line.substring(0, eq).trim();
+				String values = line.substring(eq + 1).trim();
+				
+				if(currentSection.containsKey(key)) {
+					throw new INIParseException("Section property duplicated for key \"" + key + "\" in section \"" + currentSectionName + "\".");
+				}
+				
+				currentSection.put(key, components(values));
+			}
+		}
+		
+		// If the default section is empty, remove it.
+		if(sections.get("").isEmpty()) {
+			sections.remove("");
+		}
+		
+		return sections;
+	}
+	
+	public static List<String> components (String text) throws INIParseException {
+		List<String> values = new ArrayList<String>();
+		
+		text = text.trim();
+		Scanner tokens = new Scanner(text);
+		
+		try {
+			while(tokens.hasNext()) {
+				// TODO:  Determine patterns, do tokenization, determine actual relevant values.
+				if(tokens.hasNext("")) {
+					// TODO:
+				}
+				return new ArrayList<String>();
+			}
+		} finally {
+			tokens.close();
+		}
+		
+		return values;
+	}
+	
+	static String unquote(String text) throws INIParseException {
+		// Filter both types of quotes.
+		String val = unquote_helper(text, '\'', '\'', 
+				"A quoted value was left unterminated", 
+				"Illegal text adjacent to ending quote");
+		
+		if(val == text) { // Reference equality is fine for this test, given unquote_helper's implementation.
+			return unquote_helper(val, '\"', '\"',
+				"A quoted value was left unterminated", 
+				"Illegal text adjacent to ending quote");
+		} else {
+			return val;
+		}
+
+	}
+	
+	static String unquote_helper(String text, char open, char closed, String unclosedMsg, String malformedMsg) throws INIParseException {
+		if(text.charAt(0) == open) {
+			text = text.substring(1);
+			int closeIndex = text.indexOf(closed);
+			
+			if(closeIndex == -1) {
+				throw new INIParseException(unclosedMsg + ": " + text);
+			} else if(closeIndex != text.length() - 1) {
```

Comment:
Instead, of looking for the first closing quote mark, we need to look at the last character in the (trimmed) string. For example `"a"b"c""` is considered valid by `GetPrivateProfileString` and will be returned as `a"b"c"`. Unbalanced quotes return the full string, and don't throw. We only throw at a higher level when parsing componentized strings such as in `[Files].i`